File: src/sockets/handlers/submission.handler.ts
import { Socket, Server } from 'socket.io';
import { submissionService } from '../../services/submission.service.js';
import { emitPlayerActivityUpdated } from '../emitters/room.emitter.js';
import { ServerToClientEvents, ClientToServerEvents, InterServerEvents, SocketData } from '../../types/socket.types.js';

type IOSocket = Socket<ClientToServerEvents, ServerToClientEvents, InterServerEvents, SocketData>;
type IOServer = Server<ClientToServerEvents, ServerToClientEvents, InterServerEvents, SocketData>;

export function handleSubmissionEvents(io: IOServer, socket: IOSocket) {
  const { participantId, roomId } = socket.data;

  socket.on('submission:submit', async (data, callback) => {
    try {
      await submissionService.submitChit(participantId, roomId, data.body);
      await emitPlayerActivityUpdated(io, roomId);
      callback({ success: true });
    } catch (error: any) {
      callback({ success: false, error: error.message });
    }
  });

  socket.on('submission:edit', async (data, callback) => {
    try {
      await submissionService.editChit(participantId, roomId, data.body);
      callback({ success: true });
    } catch (error: any) {
      callback({ success: false, error: error.message });
    }
  });

  socket.on('submission:delete', async (callback) => {
    try {
      await submissionService.deleteChit(participantId, roomId);
      await emitPlayerActivityUpdated(io, roomId);
      callback({ success: true });
    } catch (error: any) {
      callback({ success: false, error: error.message });
    }
  });
}

File: app/sockets/submission_handler.py
import logging
from app.core.database import AsyncSessionLocal
from app.core.logging import logger
from app.services.submission_service import (
    delete_chit,
    edit_chit,
    submit_chit,
)
from app.sockets.emitters import emit_player_activity_updated, emit_submission_status
from app.sockets.server import sio

@sio.on("submission:submit")
async def handle_submit(sid: str, data: dict):
    try:
        session = await sio.get_session(sid)
        if not session:
            return

        participant_id = session.get("participantId")
        room_id = session.get("roomId")
        body = data.get("body", "")

        async with AsyncSessionLocal() as db:
            await submit_chit(db, participant_id, room_id, body)
            await db.commit()
    except Exception as e:
        logger.error(f"Socket submit error: {e}")

@sio.on("submission:edit")
async def handle_edit(sid: str, data: dict):
    try:
        session = await sio.get_session(sid)
        if not session:
            return

        participant_id = session.get("participantId")
        room_id = session.get("roomId")
        body = data.get("body", "")

        async with AsyncSessionLocal() as db:
            await edit_chit(db, participant_id, room_id, body)
            await db.commit()
    except Exception as e:
        logger.error(f"Socket edit error: {e}")

@sio.on("submission:delete")
async def handle_delete(sid: str, data: dict = None):
    try:
        session = await sio.get_session(sid)
        if not session:
            return

        participant_id = session.get("participantId")
        room_id = session.get("roomId")

        async with AsyncSessionLocal() as db:
            await delete_chit(db, participant_id, room_id)
            await db.commit()
    except Exception as e:
        logger.error(f"Socket delete error: {e}")
